File: justbaked/src/robot_bringup/catapult/catapult_launch.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.qos import QoSProfile
from std_msgs.msg import Int32
from gpiozero import Servo
from nav2_msgs.action import NavigateToPose
from geometry_msgs.msg import PoseStamped, Quaternion
from tf2_ros import TransformException
from tf2_geometry_msgs import do_transform_pose
import yaml
import math
import time
from enum import Enum

class State(Enum):
    TO_LATCH = 1
    TURN_180 = 2
    BACK_UP = 3
    TO_HOME = 4
    FACE_FIRE = 5
    LAUNCH_CATAPULT = 6
    DONE = 7

class WaypointNavNode(Node):
    def __init__(self):
        super().__init__('waypoint_nav_node')
        self.nav_to_pose_client = ActionClient(self, NavigateToPose, 'navigate_to_pose')
        self.state = State.TO_LATCH
        self.backup_complete = False

        # Initialize poses
        self.waypoints = []
        self.latch_pose = PoseStamped()
        self.home_pose = PoseStamped()
        self.fire_pose = PoseStamped()
        self.load_waypoints()
        self.set_latch_position()
        self.home_pose.header.frame_id = 'map'
        self.home_pose.pose.position.x = 0.0
        self.home_pose.pose.position.y = 0.0
        self.home_pose.pose.position.z = 0.0
        self.home_pose.pose.orientation = self.create_quaternion(0.0)

        # Timer to execute state machine
        self.timer = self.create_timer(5.0, self.execute_state_machine)

    def load_waypoints(self):
        waypoints_file = self.get_parameter_or(
            'waypoints_file', 
            '/home/ieee/justbaked/justbaked/src/robot_bringup/waypoints/round1.yaml'
        ).get_parameter_value().string_value
        with open(waypoints_file, 'r') as file:
            config = yaml.safe_load(file)
            for wp in config['waypoints']:
                pose = PoseStamped()
                pose.header.frame_id = 'map'
                pose.pose.position.x = float(wp['x'])
                pose.pose.position.y = float(wp['y'])
                pose.pose.orientation = self.create_quaternion(float(wp['yaw']))
                self.waypoints.append(pose)
        # Assume fire waypoint is the last one (adjust as needed)
        self.fire_pose = self.waypoints[-1]
        self.get_logger().info(f"Loaded {len(self.waypoints)} waypoints")

    def set_latch_position(self):
        self.latch_pose.header.frame_id = 'map'
        self.latch_pose.pose.position.x = 0.470  # 0.470 m ahead
        self.latch_pose.pose.position.y = -0.102  # -0.102 m left
        self.latch_pose.pose.position.z = 0.0
        self.latch_pose.pose.orientation = self.create_quaternion(0.0)

    def create_quaternion(self, yaw):
        q = Quaternion()
        q.w = math.cos(yaw / 2.0)
        q.x = 0.0
        q.y = 0.0
        q.z = math.sin(yaw / 2.0)
        return q

    def execute_state_machine(self):
        if not self.nav_to_pose_client.wait_for_server(timeout_sec=5.0):
            self.get_logger().error("Nav2 action server not available")
            return

        if self.state == State.TO_LATCH:
            self.get_logger().info("Navigating to just in front of latch...")
            self.send_goal(self.latch_pose, self.to_latch_callback)

        elif self.state == State.TURN_180:
            self.get_logger().info("Turning 180 degrees...")
            self.latch_pose.pose.orientation = self.turn_180(self.latch_pose.pose.orientation)
            self.send_goal(self.latch_pose, self.turn_180_callback)

        elif self.state == State.BACK_UP:
            self.get_logger().info("Backing up several inches...")
            self.backup_robot()

        elif self.state == State.TO_HOME:
            self.get_logger().info("Navigating back to origin...")
            self.send_goal(self.home_pose, self.to_home_callback)

        elif self.state == State.FACE_FIRE:
            self.get_logger().info("Facing the fire waypoint...")
            self.home_pose.pose.orientation = self.face_position(self.fire_pose.pose.position)
            self.send_goal(self.home_pose, self.face_fire_callback)

        elif self.state == State.LAUNCH_CATAPULT:
            self.get_logger().info("Launching catapult...")
            self.launch_catapult()
            self.state = State.DONE

        elif self.state == State.DONE:
            self.get_logger().info("Task complete.")
            self.timer.cancel()

    def send_goal(self, pose, result_callback):
        goal_msg = NavigateToPose.Goal()
        goal_msg.pose = pose
        goal_handle_future = self.nav_to_pose_client.send_goal_async(
            goal_msg,
            feedback_callback=None
        )
        goal_handle_future.add_done_callback(
            lambda future: self.goal_response_callback(future, result_callback)
        )

    def goal_response_callback(self, future, result_callback):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error("Goal rejected")
            return
        self.get_logger().info("Goal accepted")
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(result_callback)

    def to_latch_callback(self, future):
        result = future.result().result
        if result and result.result == NavigateToPose.Result().SUCCESS:
            self.state = State.TURN_180

    def turn_180_callback(self, future):
        result = future.result().result
        if result and result.result == NavigateToPose.Result().SUCCESS:
            self.state = State.BACK_UP

    def turn_180(self, current):
        yaw = 2 * math.acos(current.w)  # Extract current yaw
        new_yaw = yaw + math.pi  # Add 180 degrees (π radians)
        return self.create_quaternion(new_yaw)

    def backup_robot(self):
        backup_pose = self.latch_pose
        backup_pose.pose.position.x -= 0.1524  # Back up 6 inches
        self.send_goal(backup_pose, self.backup_callback)

    def backup_callback(self, future):
        result = future.result().result
        if result and result.result == NavigateToPose.Result().SUCCESS:
            time.sleep(2)  # Wait 2 seconds
            self.backup_complete = True
            self.state = State.TO_HOME

    def to_home_callback(self, future):
        result = future.result().result
        if result and result.result == NavigateToPose.Result().SUCCESS:
            self.state = State.FACE_FIRE

    def face_fire_callback(self, future):
        result = future.result().result
        if result and result.result == NavigateToPose.Result().SUCCESS:
            self.state = State.LAUNCH_CATAPULT

    def face_position(self, target):
        dx = target.x - self.home_pose.pose.position.x
        dy = target.y - self.home_pose.pose.position.y
        yaw = math.atan2(dy, dx)
        return self.create_quaternion(yaw)

    def launch_catapult(self):
        # Replace with your servo motor control code
        self.get_logger().info("Servo motor activating catapult...")
        # Set up the servo (GPIO 5 aka Pin 29)
        self.servo = Servo(5)
        
        # Create a subscription to listen to angle values
        self.subscription = self.create_subscription(
                Int32,
                'servo_angle', # Topic name
                self.listener_callback,
                10
        )
        self.subscription # prevent unused variable warning

        angle = msg.data # Get angle from the message
        self.get_logger().info(f"Moving servo to: {angle} degrees")

        # Map the angle to the servo's range of motion (-1 to 1) (Can be adjusted based on servo)
        mapped_value = (angle - 90) / 90 # Map angle (90 degrees as neutral)

        # Move the servo to the mapped position
        self.servo.value = mapped_value
        time.sleep(1)


def main(args=None):
    rclpy.init(args=args)
    node = WaypointNavNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: justbaked/src/robot_bringup/catapult/catapult_launch.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <nav2_msgs/action/navigate_to_pose.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <std_msgs/msg/int32.hpp>
#include <yaml-cpp/yaml.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <math.h>
#include <string>
#include <vector>
#include <chrono>

using namespace std::chrono_literals;

class CatapultNode : public rclcpp::Node {
public:
  enum class State {
    TO_LATCH = 1,
    TURN_180 = 2,
    BACK_UP = 3,
    TO_HOME = 4,
    FACE_FIRE = 5,
    LAUNCH_CATAPULT = 6,
    DONE = 7
  };

  CatapultNode() : Node("catapult_node") {
    nav_to_pose_client_ = rclcpp_action::create_client<nav2_msgs::action::NavigateToPose>(
      this, "navigate_to_pose");

    // Initialize state and other parameters
    state_ = State::TO_LATCH;
    backup_complete_ = false;

    // Initialize poses
    latch_pose_ = geometry_msgs::msg::PoseStamped();
    home_pose_ = geometry_msgs::msg::PoseStamped();
    fire_pose_ = geometry_msgs::msg::PoseStamped();
    load_waypoints();
    set_latch_position();
    home_pose_.header.frame_id = "map";
    home_pose_.pose.position.x = 0.0;
    home_pose_.pose.position.y = 0.0;
    home_pose_.pose.position.z = 0.0;
    home_pose_.pose.orientation = create_quaternion(0.0);

    // Publisher for the servo
    servo_pub_ = this->create_publisher<std_msgs::msg::Int32>("servo_angle", 10);

    // Timer for the state machine
    timer_ = this->create_wall_timer(5s, std::bind(&CatapultNode::execute_state_machine, this));
  }

private:
  rclcpp_action::Client<nav2_msgs::action::NavigateToPose>::SharedPtr nav_to_pose_client_;
  rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr servo_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::vector<geometry_msgs::msg::PoseStamped> waypoints_;
  geometry_msgs::msg::PoseStamped latch_pose_;
  geometry_msgs::msg::PoseStamped home_pose_;
  geometry_msgs::msg::PoseStamped fire_pose_;
  State state_;
  bool backup_complete_;

  void load_waypoints() {
    std::string waypoints_file = this->declare_parameter<std::string>(
      "waypoints_file", "/home/ieee/justbaked/justbaked/src/robot_bringup/waypoints/round3.yaml");

    YAML::Node config = YAML::LoadFile(waypoints_file);
    for (const auto &wp : config["waypoints"]) {
      geometry_msgs::msg::PoseStamped pose;
      pose.header.frame_id = "map";
      pose.pose.position.x = wp["x"].as<double>();
      pose.pose.position.y = wp["y"].as<double>();
      double yaw = wp["yaw"].as<double>();
      pose.pose.orientation = create_quaternion(yaw);
      waypoints_.push_back(pose);
    }

    // Assuming fire waypoint is the last one (adjust as needed)
    fire_pose_ = waypoints_.back();
    RCLCPP_INFO(this->get_logger(), "Loaded %zu waypoints", waypoints_.size());
  }

  void set_latch_position() {
    latch_pose_.header.frame_id = "map";
    latch_pose_.pose.position.x = 1.524 - 0.349;  // 5 ft - 13.5" ahead
    latch_pose_.pose.position.y = 0.457 - 0.349;  // 1.5 ft - 11.5" to the side
    latch_pose_.pose.position.z = 0.0;
    latch_pose_.pose.orientation = create_quaternion(M_PI);  // 180 degrees yaw
  }

  geometry_msgs::msg::Quaternion create_quaternion(double yaw) {
    geometry_msgs::msg::Quaternion q;
    q.w = std::cos(yaw / 2.0);
    q.x = 0.0;
    q.y = 0.0;
    q.z = std::sin(yaw / 2.0);
    return q;
  }

  void execute_state_machine() {
    if (!nav_to_pose_client_->wait_for_action_server(5s)) {
      RCLCPP_ERROR(this->get_logger(), "Nav2 action server not available");
      return;
    }

    switch (state_) {
      case State::TO_LATCH:
        RCLCPP_INFO(this->get_logger(), "Navigating to just in front of latch...");
        send_goal(latch_pose_, std::bind(&CatapultNode::to_latch_callback, this, std::placeholders::_1));
        break;

      case State::TURN_180:
        RCLCPP_INFO(this->get_logger(), "Turning 180 degrees...");
        latch_pose_.pose.orientation = turn_180(latch_pose_.pose.orientation);
        send_goal(latch_pose_, std::bind(&CatapultNode::turn_180_callback, this, std::placeholders::_1));
        break;

      case State::BACK_UP:
        RCLCPP_INFO(this->get_logger(), "Backing up...");
        backup_robot();
        break;

      case State::TO_HOME:
        RCLCPP_INFO(this->get_logger(), "Navigating back to home...");
        send_goal(home_pose_, std::bind(&CatapultNode::to_home_callback, this, std::placeholders::_1));
        break;

      case State::FACE_FIRE:
        RCLCPP_INFO(this->get_logger(), "Facing the fire waypoint...");
        home_pose_.pose.orientation = face_position(fire_pose_.pose.position);
        send_goal(home_pose_, std::bind(&CatapultNode::face_fire_callback, this, std::placeholders::_1));
        break;

      case State::LAUNCH_CATAPULT:
        RCLCPP_INFO(this->get_logger(), "Launching catapult...");
        launch_catapult();
        state_ = State::DONE;
        break;

      case State::DONE:
        RCLCPP_INFO(this->get_logger(), "Task complete.");
        timer_->cancel();
        break;
    }
  }

  void send_goal(geometry_msgs::msg::PoseStamped pose, std::function<void(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr)> result_callback) {
    auto goal_msg = nav2_msgs::action::NavigateToPose::Goal();
    goal_msg.pose = pose;

    auto goal_handle_future = nav_to_pose_client_->async_send_goal(goal_msg);
    goal_handle_future.add_done_callback(
      [this, result_callback](std::future<rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr> future) {
        auto goal_handle = future.get();
        if (goal_handle) {
          result_callback(goal_handle);
        } else {
          RCLCPP_ERROR(this->get_logger(), "Goal rejected");
        }
      }
    );
  }

  void to_latch_callback(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr goal_handle) {
    state_ = State::TURN_180;
  }

  void turn_180_callback(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr goal_handle) {
    state_ = State::BACK_UP;
  }

  void backup_robot() {
    latch_pose_.pose.position.x -= 0.1524;  // Back up 6 inches
    send_goal(latch_pose_, std::bind(&CatapultNode::backup_callback, this, std::placeholders::_1));
  }

  void backup_callback(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr goal_handle) {
    backup_complete_ = true;
    state_ = State::TO_HOME;
  }

  void to_home_callback(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr goal_handle) {
    state_ = State::FACE_FIRE;
  }

  void face_fire_callback(rclcpp_action::ClientGoalHandle<nav2_msgs::action::NavigateToPose>::SharedPtr goal_handle) {
    state_ = State::LAUNCH_CATAPULT;
  }

  geometry_msgs::msg::Quaternion turn_180(geometry_msgs::msg::Quaternion current) {
    double yaw = 2 * std::acos(current.w);  // Extract current yaw
    double new_yaw = yaw + M_PI;  // Add 180 degrees (π radians)
    return create_quaternion(new_yaw);
  }

  geometry_msgs::msg::Quaternion face_position(const geometry_msgs::msg::Point &target) {
    double dx = target.x - home_pose_.pose.position.x;
    double dy = target.y - home_pose_.pose.position.y;
    double yaw = std::atan2(dy, dx);
    return create_quaternion(yaw);
  }

  void launch_catapult() {
    // Send 36 degrees to trigger the catapult launch
    std_msgs::msg::Int32 msg;
    msg.data = 36;  // Set the angle to 36 degrees for the catapult launch
    servo_pub_->publish(msg);
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CatapultNode>());
  rclcpp::shutdown();
  return 0;
}

